libs/utils/src/properties.c

```diff
@@ -191,324 +476,423 @@ static void parseLine(const char* line, celix_properties_t *props) {
-                }
-                filebuffer[file_size]='\0';
-                line = strtok_r(filebuffer, "\n", &saveptr);
-                while (line != NULL) {
-                    parseLine(line, props);
-                    line = strtok_r(NULL, "\n", &saveptr);
-                }
-                free(filebuffer);
+/**
+ * @brief Store properties string to file and escape the characters '#', '!', '=' and ':' if encountered.
+ */
+static int celix_properties_storeEscapedString(FILE* file, const char* str) {
+    int rc = 0;
+    for (int i = 0; i < strlen(str); i += 1) {
+        if (str[i] == '#' || str[i] == '!' || str[i] == '=' || str[i] == ':') {
+            rc = fputc('\\', file);
+            if (rc == EOF) {
+                break;
             }
         }
+        rc = fputc(str[i], file);
     }
-
-    return props;
+    return rc;
 }
 
-celix_properties_t* celix_properties_loadFromString(const char *input) {
-    celix_properties_t *props = celix_properties_create();
+celix_status_t celix_properties_store(celix_properties_t* properties, const char* filename, const char* header) {
+    FILE* file = fopen(filename, "w+");
+
+    if (file == NULL) {
+        celix_err_pushf("Cannot open file '%s'", filename);
+        return CELIX_FILE_IO_EXCEPTION;
+    }
 
-    char *in = strdup(input);
-    char *line = NULL;
-    char *saveLinePointer = NULL;
+    int rc = 0;
 
-    bool firstTime = true;
-    do {
-        if (firstTime){
-            line = strtok_r(in, "\n", &saveLinePointer);
-            firstTime = false;
-        }else {
-            line = strtok_r(NULL, "\n", &saveLinePointer);
+    if (header && strstr("\n", header)) {
+        celix_err_push("Header cannot contain newlines. Ignoring header.");
+    } else if (header) {
+        rc = fputc('#', file);
+        if (rc != 0) {
+            rc = fputs(header, file);
         }
-
-        if (line == NULL){
-            break;
+        if (rc != 0) {
+            rc = fputc('\n', file);
         }
+    }
 
-        parseLine(line, props);
-    } while(line != NULL);
+    CELIX_PROPERTIES_ITERATE(properties, iter) {
+        const char* val = iter.entry.value;
+        if (rc != EOF) {
+            rc = celix_properties_storeEscapedString(file, iter.key);
+        }
+        if (rc != EOF) {
+            rc = fputc('=', file);
+        }
+        if (rc != EOF) {
+            rc = celix_properties_storeEscapedString(file, val);
+        }
+        if (rc != EOF) {
+            rc = fputc('\n', file);
+        }
+    }
+    if (rc != EOF) {
+        rc = fclose(file);
+    } else {
+        fclose(file);
+    }
+    if (rc == EOF) {
+        celix_err_push("Failed to write properties to file");
+        return CELIX_FILE_IO_EXCEPTION;
+    }
+    return CELIX_SUCCESS;
+}
 
-    free(in);
+celix_properties_t* celix_properties_copy(const celix_properties_t* properties) {
+    celix_properties_t* copy = celix_properties_create();
```

Comment:
If any of the copy failed, we'd better cleanup and return null. Dropping property silently is not a good idea.